presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -0,0 +1,440 @@
+import com.facebook.presto.spi.type.ParameterKind;
+import com.facebook.presto.spi.type.StandardTypes;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.spi.type.TypeSignatureParameter;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Sets;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.facebook.presto.type.TypeCalculation.calculateLiteralValue;
+import static com.facebook.presto.type.TypeRegistry.canCoerce;
+import static com.facebook.presto.type.TypeRegistry.getCommonSuperTypeSignature;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterables.getLast;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+public class SignatureBinder
+{
+    private final TypeManager typeManager;
+    private final Signature declaredSignature;
+    private final boolean allowCoercion;
+    private final Map<String, TypeVariableConstraint> typeVariableConstraints;
+
+    public SignatureBinder(TypeManager typeManager, Signature declaredSignature, boolean allowCoercion)
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.declaredSignature = requireNonNull(declaredSignature, "parametrizedSignature is null");
+        this.allowCoercion = allowCoercion;
+        this.typeVariableConstraints = declaredSignature.getTypeVariableConstraints()
+                .stream()
+                .collect(toMap(TypeVariableConstraint::getName, t -> t));
+    }
+
+    public Optional<Signature> matchAndBindSignature(List<? extends Type> actualArgumentTypes)
+    {
+        Optional<BoundVariables> boundVariables = matchAndBindSignatureVariables(actualArgumentTypes);
+        if (!boundVariables.isPresent()) {
+            return Optional.empty();
+        }
+        return Optional.of(bindVariables(declaredSignature, boundVariables.get(), actualArgumentTypes.size()));
+    }
+
+    public Optional<BoundVariables> matchAndBindSignatureVariables(List<? extends Type> actualArgumentTypes)
+    {
+        List<TypeSignature> expectedArgumentSignatures = declaredSignature.getArgumentTypes();
+        boolean variableArity = declaredSignature.isVariableArity();
+        BoundVariables.Builder boundVariablesBuilder = BoundVariables.builder();
+        if (!matchArguments(expectedArgumentSignatures, actualArgumentTypes, variableArity, boundVariablesBuilder)) {
+            return Optional.empty();
+        }
+        calculateVariableValuesForLongConstraints(boundVariablesBuilder);
+        BoundVariables boundVariables = boundVariablesBuilder.build();
+        checkAllTypeVariablesAreBound(boundVariables);
```

Comment:
Than we need to pass required variables list to the builder. Do you think it is better than calling this method?